crates/smtp_session/src/message.rs

```diff
@@ -0,0 +1,150 @@
+use error_handler::Error;
+
+#[derive(Debug)]
+pub struct SmtpMessage {
+    pub from: String,
+    pub to: Vec<String>,
+    pub subject: String,
+    pub body: String,
+}
+
+impl SmtpMessage {
+
+    pub fn new() -> SmtpMessageBuilder {
+        SmtpMessageBuilder::default()
+    }
+
+    pub fn to_imf(&self) -> String {
+        let mut imf_message = String::new();
+        
+        imf_message.push_str(&format!("From: {}\r\n", self.from));
+        
+        imf_message.push_str(&format!("To: {}\r\n", self.to.join(", ") ));
+
+        imf_message.push_str(&format!("Subject: {}\r\n", self.subject));
+        
+        imf_message.push_str("\r\n");
+        
+        imf_message.push_str(&self.body);
+        
+        imf_message
+    }
+}
+
+
+
+#[derive(Default)]
+pub struct SmtpMessageBuilder {
+    from: Option<String>,
+    to: Vec<String>,
+    subject: Option<String>,
+    body: Option<String>,
+}
+
+impl SmtpMessageBuilder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn from(mut self, from: &str) -> Self {
+        self.from = Some(from.to_string());
+        self
+    }
+
+    pub fn to(mut self, to: &str) -> Self {
+        self.to.push(to.to_string());
+        self
+    }
+
+    pub fn subject(mut self, subject: &str) -> Self {
+        self.subject = Some(subject.to_string());
+        self
+    }
+
+    pub fn body(mut self, body: &str) -> Self {
+        self.body = Some(body.to_string());
+        self
+    }
+
+    pub fn build(self) -> Result<SmtpMessage, Error> {
+        if self.from.is_none() {
+            return Err(Error::MessageBuild("Missing 'from' field".to_string()));
+        }
+
+        if self.to.is_empty() {
+            return Err(Error::MessageBuild("Missing 'to' field".to_string()));
+        }
+
+        if self.subject.is_none() {
+            return Err(Error::MessageBuild("Missing 'subject' field".to_string()));
+        }
+
+        if self.body.is_none() {
+            return Err(Error::MessageBuild("Missing 'body' field".to_string()));
+        }
+
+        Ok(SmtpMessage {
+            from: self.from.unwrap(),
+            to: self.to,
+            subject: self.subject.unwrap(),
+            body: self.body.unwrap(),
+        })
+    }
+}
+
+
+
+
+// tests
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_smtp_to_imf() {
+        let message = SmtpMessage::new()
```

Comment:
By changing the constructor you will end up with
```
let message = SmtpMessage::new(
    "johndoe@gmail.com",
    "emilydoe@gmail.com",
    "Hello",
    "Hello, Emily!"));
```
or if you want extra readability
```
let sender = "johndoe@gmail.com";
let recipient = "emilydoe@gmail.com";
let subject = "Hello";
let body = "Hello, Emily!";
let message = SmtpMessage::new(sender, recipient, subject, body);
```